File: Synapse_Load_and_Save.py
#!/usr/bin/env python2
# -*- coding: utf-8 -*-
"""
Created on Tue May 22 10:47:54 2018

@author: cl321
"""

import synapseclient
import argparse
import os
import datetime
import pandas as pd
import sys
import shutil

"""get user input"""
parser = argparse.ArgumentParser()
parser.add_argument('username')
parser.add_argument('password')
parser.add_argument('dir_project')
parser.add_argument('synapse_ids')
parser.add_argument('R_variable_name', nargs='?', default=None)

args = parser.parse_args()

username = args.username
password = args.password
dir_project = args.dir_project
synapse_ids = args.synapse_ids
R_variable_name = args.R_variable_name

if R_variable_name != None:
    R_variable_name = R_variable_name.split(',')
synapse_id_list = synapse_ids.split(',')
    
"""make connect to synapse, set directories"""
syn=synapseclient.Synapse()
syn.login(username, password)

dir_c_script = os.getcwd()
dir_log = dir_c_script + '/' + os.path.basename(dir_c_script)+'_SynapseLog'
time = datetime.datetime.now().strftime('%Y%m%d_%H%M%S')
tempdir = dir_project + 'temp_local_synapse_files'+time
os.mkdir(tempdir)

def get_synapse_info(synapse_id, download):
    file_info = {}
    for index, syn_id in enumerate(synapse_id_list):
        entity = syn.get(syn_id, downloadFile=download, downloadLocation=tempdir,
                         ifcollision= 'overwrite.local')
        file_info[index]={}
        try:
            file_info[index]['file_name'] = entity.name
        except AttributeError:
            pass
        if R_variable_name==None:
            file_info[index]['R_variable_name'] = os.path.splitext(entity.name)[0]
        else:
            file_info[index]['R_variable_name'] = R_variable_name[index]
        try:
            file_info[index]['id'] = entity.id
            file_info[index]['url']='http://synapse.sagebase.org/#Synapse:'+entity.id
        except AttributeError:
            pass
        try:
            file_info[index]['description'] = entity.description
        except AttributeError:
            pass
        try:
            file_info[index]['url'] = entity.url
        except AttributeError:
            pass
        try:
            file_info[index]['version_number'] = entity.versionNumber
        except AttributeError:
            pass
        try:
            file_info[index]['versions_label'] = entity.versionLabel
        except AttributeError:
            pass
        try:
            file_info[index]['DateTime_modified'] = entity.modifiedOn
        except AttributeError:
            pass
    file_info_final = pd.DataFrame(file_info).T
    return file_info_final

"""get info of files copied in for user to check"""
check_file_info = get_synapse_info(synapse_id_list, False)

print 'to be loaded '+ ','.join(check_file_info['file_name'].tolist())
print '==========================================================='
print "please check 'file_info' before continuing loading process"

"""download files from Synapse"""
file_info_tosave = get_synapse_info(synapse_id_list, True)

syn.logout()

"""write logfile"""
if not os.path.exists(dir_log):
    os.mkdir(dir_log)

currentscript_insert = os.path.splitext(os.path.basename(sys.argv[0]))[0]
[logfilename1, logfilename2] = datetime.datetime.now().strftime('%Y%m%d_%H%M%S').split('_')
logfile_name = dir_log+'/FilesUsed_'+logfilename1+'_'+currentscript_insert+'_'+logfilename2+'_GMT.csv'
file_info_tosave.to_csv(logfile_name, index=False)
    
"""load data into R"""
files = file_info_tosave['file_name'].tolist()
if len(files)==0:    
    print 'something went wrong'

filedict = {}
for index, afile in enumerate(files):
    varname = file_info_tosave.iloc[index]['R_variable_name']
    afile = pd.read_csv(tempdir+'/'+file_info_tosave.iloc[index]['file_name'],
                        index_col=0)
    filedict[varname] = afile

shutil.rmtree(tempdir)    

File: Synapse_Load_and_Save.R
#source('http://depot.sagebase.org/CRAN.R')
#pkgInstall("synapseClient")
#install.packages("rstudioapi")

library(synapseClient)
library(plyr);library(dplyr)
library(tidyr)
library(ggplot2)
library(ggrepel)
`%nin%`<-Negate(`%in%`)


##! to do: divide this script into a) create database table, b) create figures

##################################################################################################################T
# Get user input ------------
##################################################################################################################T
#User input
Synapse_UID<-""
Synapse_password<-""
dir_project<-""
synapse_ids<-c("syn11701291", "syn12617467","syn13549817")#c("synID_1","synID_2",...,"synID_n") ## requires user input
variable_names<-NA#optional: change to c("myname_1","myname_2",...,"myname_n") ## optional user input

##################################################################################################################T
# make connection to synapse, set directories ------------
##################################################################################################################T
synapseLogin(Synapse_UID, Synapse_password) ## requires user input
c.time<-Sys.time()%>%as.character()%>%gsub("-","",.)%>%gsub(" ","_",.)%>%gsub(":","",.)
#dir_project<-"/Users/nienke/Dropbox (HMS)/CBDM-SORGER/Collaborations/Dark_kinome/" ##requires user input
paste0(dir_project,paste0(rstudioapi::getActiveDocumentContext()$path%>%dirname()%>%basename()))
dir_c_script<-paste0(dir_project,paste0(rstudioapi::getActiveDocumentContext()$path%>%dirname()%>%basename()))
dir_log<-paste0(dir_project,paste0(rstudioapi::getActiveDocumentContext()$path%>%dirname()%>%basename()),"_SynapseLog")
tempdir<-paste0(dir_project,"temp_local_synapse_files",Sys.time()%>%as.character()%>%gsub("-","",.)%>%gsub(" ","_",.)%>%gsub(":","",.))
dir.create(tempdir, showWarnings = TRUE, recursive = TRUE)
#synapse_ids<-c("syn12074621","syn12074611")#c("synID_1","synID_2",...,"synID_n") ## requires user input
#variable_names<-NA#optional: change to c("myname_1","myname_2",...,"myname_n") ## optional user input
##################################################################################################################T
# get info of files copied in for user to check ------------
##################################################################################################################T
file_info.l<-list()
i=0
for(syn_id in synapse_ids){
  i=i+1
  syn_file<-synGet(syn_id, downloadFile =F, downloadLocation = tempdir, ifcollision = "overwrite.local")
  c.info<-list()
  c.info$file_name<-syn_file[[1]]$name
  c.info$R_variable_name<-ifelse(is.na(variable_names)==T,tools::file_path_sans_ext(c.info$file_name),variable_names[i])
  c.info$id<-syn_file[[1]]$id%>%as.character(.)
  c.info$descreption<-syn_file[[1]]$description
  c.info$url<-syn_file@synapseWebUrl
  c.info$version_number<-syn_file[[1]]$versionNumber
  c.info$versions_label<-syn_file[[1]]$versionLabel
  c.info$DateTime_modified<-syn_file[[1]]$modifiedOn
  file_info.l[[i]]<-c.info%>%as.data.frame(.,stringsAsFactors=F)
}
file_info<-file_info.l%>%bind_rows(.)

writeLines(
  paste0("to be loaded:\n ",
         file_info$file_name%>%toString()%>%gsub(",","\n",.),
         "\n===========================================================
         \nplease check 'file_info' before continuing loading process"))

##################################################################################################################T
# download files from Synapse ------------
##################################################################################################################T

for(syn_id in synapse_ids){
  syn_file<-synGet(syn_id, downloadFile =T, downloadLocation = tempdir, ifcollision = "overwrite.local")
  c.info<-list()
  c.info$file_name<-syn_file[[1]]$name
  c.info$R_variable_name<-ifelse(is.na(variable_names)==T,tools::file_path_sans_ext(c.info$file_name),variable_names[i])
  c.info$id<-syn_file[[1]]$id
  c.info$descreption<-syn_file[[1]]$description
  c.info$url<-syn_file@synapseWebUrl
  c.info$version_number<-syn_file[[1]]$versionNumber
  c.info$versions_label<-syn_file[[1]]$versionLabel
  c.info$DateTime_modified<-syn_file[[1]]$modifiedOn
  file_info.l[[i]]<-c.info%>%as.data.frame(.,stringsAsFactors=F)
}
file_info<-file_info.l%>%bind_rows(.)
synapseLogout()

##################################################################################################################T
# write logfile ------------
##################################################################################################################T
setwd(dir_c_script)
paste0(dir_log)%nin%list.files()
if(dir_log%nin%list.files()){
  dir.create(dir_log, showWarnings = TRUE, recursive = TRUE)}

currentscript_insert<-paste0("_",
                             rstudioapi::getActiveDocumentContext()$path%>%basename()%>%gsub(".R","",.),
                             "_")
logfile_name<-Sys.time()%>%as.character()%>%gsub("-","",.)%>%
  gsub(" ",currentscript_insert,.)%>%gsub(":","",.)%>%paste0(.,"GMT")%>%
  paste0("FilesUsed_",.,".csv")

#Sys.time()%>%as.character()%>%gsub("-","",.)%>%
#  gsub(" ",currentscript_insert,.)%>%gsub(":","",.)%>%paste0(.,"GMT")%>%
#  paste0("FilesUsed_",.,".csv")

setwd(dir_log)
write.csv(file_info,file = logfile_name,row.names = F)

##################################################################################################################T
# load data into R ------------
##################################################################################################################T
##!!modify to allow more datatypes
setwd(tempdir)
file_info<-file_info%>%arrange(file_name)
files<-ifelse(list.files()%>%sort()==file_info$file_name%>%as.character(),
              list.files(),"something went wrong")
i=0
for(c.file in files){
  i=i+1
  c.varname<-file_info$R_variable_name[i]%>%as.character()
  c.file<-read.csv(c.file, stringsAsFactors = F) #!!modify to allow more datatypes
  assign(c.varname,c.file)
}
##################################################################################################################T
# remove excess variables ------------
##################################################################################################################T
vars_to_keep<-c(c("%nin%","file_info",file_info$R_variable_name),
                "Synapse_UID","Synapse_password", "dir_project","tempdir",
                "dir_log","currentscript_insert","dir_c_script")
rm(list = ls()[ls()%nin%vars_to_keep])

##################################################################################################################T
# Manipulate data (add more headers if needed) ------------
##################################################################################################################T

# customize per dataset

##################################################################################################################T
# user input for upload  ------------
##################################################################################################################T
file_info_upload<-
  data.frame(Rvariable_name=c("Phospho_baseline_g_day"),
             filenames=c("Table_006_RenMVcell_phosphoproteomics.csv"), 
             file_description=c("Summary table phosphoproteomics of RenMVcell differentiation 15 days"),
             synapse_folder_id=c("syn12176087"),
             synapse_folder_url=c("https://www.synapse.org/Portal.html#!Synapse:syn12176087"),
             date_time=c(Sys.time()%>%as.character()%>%gsub("-","",.)%>%paste0(.,"GMT")),
             Rscript=c(currentscript_insert),
             local_copy_name=c("phosphoproteomics_RenMVcellDifferentiation_databasetable.csv"),
             stringsAsFactors = F)
#synapse_folder<-"syn12176087"

##################################################################################################################T
# create logfile input  ------------
##################################################################################################################T
logfile_name_upload<-Sys.time()%>%as.character()%>%gsub("-","",.)%>%
  gsub(" ",currentscript_insert,.)%>%gsub(":","",.)%>%paste0(.,"GMT")%>%
  paste0("FilesUploaded_",.,".csv")
setwd(dir_log)
write.csv(file_info_upload,logfile_name_upload,row.names = F)

##################################################################################################################T
# upload to synapse  ------------
##################################################################################################################T
synapseLogin(Synapse_UID, Synapse_password)
setwd(tempdir)
for(i in 1:dim(file_info_upload)[1]){
  write.csv(get(file_info_upload$Rvariable_name[i]),file=file_info_upload$filenames[i],row.names = F)
  c.file<-File(paste0(tempdir,"/",file_info_upload$filenames[i]),
               parentId=file_info_upload$synapse_folder_id[i], synapseStore = T)
  synSetAnnotations(c.file)<-list(description = file_info_upload$file_description[i])
  file<-synStore(c.file)
  print(paste0(i,"-",dim(file_info_upload)[1]))
}
synapseLogout()
##################################################################################################################T
# remove tempdir  ------------
##################################################################################################################T
setwd(dir_project)
unlink(tempdir,recursive =T)



